[BE][Team-15, yeon] baseball 기능 구현과 배포

kimnayeon0108 opened the pull request:
안녕하세요 백엔드 yeon입니다!
기능 구현과 ec2 인스턴스 두대로 나눠서 웹서버와 DB를 따로 둬서 배포하는 작업까지 완료했습니다.
프론트엔드 분들이 cors설정 헤더를 추가해달라고 요청하여, CorsFilter라는 클래스를 추가하여 Response에 헤더를 추가하는 방식으로 진행했습니다. 
OAuth는 아직 구현하지 않았고 곧 진행할 예정입니다.

[배포 URL](http://ec2-13-125-90-225.ap-northeast-2.compute.amazonaws.com/teams)

**DB설계**
<img width="630" alt="스크린샷 2021-05-11 오후 4 39 15" src="https://user-images.githubusercontent.com/65011131/117777326-75138d00-b277-11eb-8d60-e4e0d1cda1ba.png">

**[API 설계](https://github.com/kimnayeon0108/baseball/wiki/baseball-API-%EC%84%A4%EA%B3%84)**

시간내어 리뷰 해주셔서 감사합니다.

kimnayeon0108 committed:
Chore: Add BE, FE directory and .gitignore

kimnayeon0108 committed:
Feat: add configuration of JDBC connection with mysql and spring boot project

kimnayeon0108 committed:
Feat: add Team domain class

kimnayeon0108 committed:
Feat: add Team domain class

kimnayeon0108 committed:
Feat: add TeamRepository dao class

kimnayeon0108 committed:
Feat: add RequestTeamDTO class to add team data into database

kimnayeon0108 committed:
Feat: add TeamController, inserting team data function

kimnayeon0108 committed:
Feat: add schema.sql that create tables

kimnayeon0108 committed:
fixed untracked files

kimnayeon0108 committed:
Chore: update .gitignore

kimnayeon0108 committed:
Chore: update .gitignore

kimnayeon0108 committed:
Chore: update .gitignore

kimnayeon0108 committed:
Chore: update .gitignore

kimnayeon0108 committed:
Chore: fixed untracked files

kimnayeon0108 committed:
Feat: add scores to have Scores in Team

kimnayeon0108 committed:
Feat: add Record domain class for record table

kimnayeon0108 committed:
Feat: add Record field in Member class

kimnayeon0108 committed:
Feat: add RecordMember class for record table joined with member table

kimnayeon0108 committed:
Feat: add joining record table with member table function

kimnayeon0108 committed:
Feat: add MemberDTO and TeamDTO for showing team and member without record of each member

kimnayeon0108 committed:
Feat: add TeamRecordsDTO for showing records including teamId

kimnayeon0108 committed:
Feat: add showing teams and records of team members function

kimnayeon0108 committed:
Feat: add Game domain class for game table

kimnayeon0108 committed:
Feat: add Score domain class for score table

kimnayeon0108 committed:
Feat: add GameRepository as DAO

kimnayeon0108 committed:
Feat: update schema.sql

kimnayeon0108 committed:
Feat: add GameDTO class

kimnayeon0108 committed:
Feat: add ScoredDTO class

kimnayeon0108 committed:
Feat: add showing games and scores of teams

kimnayeon0108 committed:
Feat: add TeamService to move business logic from controller

kimnayeon0108 committed:
Feat: move business logic to TeamService

kimnayeon0108 committed:
Chore: code align

kimnayeon0108 committed:
Feat: add convert members set to MemberDTO list function

kimnayeon0108 committed:
Chore: remove useless import statement

kimnayeon0108 committed:
Refactor: remove useless method

kimnayeon0108 committed:
Feat: add function to get to know whether member has record or not

kimnayeon0108 committed:
Feat: add getting Member by member id through Team object

kimnayeon0108 committed:
Refactor: change DTO name, RecordMember to RecordDTO

kimnayeon0108 committed:
Refactor: remove useless factory method

kimnayeon0108 committed:
Refactor: change DTO name, RecordMember to RecordDTO

kimnayeon0108 committed:
Feat: add function to add score of team

kimnayeon0108 committed:
Refactor: replace ScoreDTOs with TeamScoreDTO

kimnayeon0108 committed:
Feat: add GameScoreDTO to include gameId

kimnayeon0108 committed:
Refactor: replace RecordMember with RecordDTO

kimnayeon0108 committed:
Feature: add RequestScoreDTO for request body

kimnayeon0108 committed:
Feature: add RequestRecordDTO for request body

kimnayeon0108 committed:
Feature: move business logic to GameService from controller

kimnayeon0108 committed:
Refactor: replace int type parameters with RequestRecordDTO

kimnayeon0108 committed:
Feat: add custom exception

kimnayeon0108 committed:
Refactor: replace exception with custom exception

kimnayeon0108 committed:
Feat: add exception handler

kimnayeon0108 committed:
Feat: update schema.sql

kimnayeon0108 committed:
Feat: update mysql connection url

kimnayeon0108 committed:
Test: remove test file

kimnayeon0108 committed:
Refactor: update jdbc connection url

kimnayeon0108 committed:
Feat: add CorsFilter to add header for cors

ksundong commented on BE/src/main/java/baseball/BaseballApplication.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.web.servlet.ServletComponentScan;
+
+@ServletComponentScan
```
오호 저도 처음 알게된 애노테이션이네요. 😮

kimnayeon0108 replied:
헤더 추가하는 법 검색하다가 블로그에 나온대로 사용했습니다. 어노테이션 없애봤는데 잘 동작하네요...

ksundong commented on BE/src/main/java/baseball/CorsFilter.java:
```diff
@@ -0,0 +1,29 @@
+@Component
+public class CorsFilter implements Filter {
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletResponse httpServletResponse = (HttpServletResponse) response;
+        httpServletResponse.setHeader("Access-Control-Allow-Origin", "*");
+        httpServletResponse.setHeader("Access-Control-Allow-Credentials", "true");
+        chain.doFilter(request, response);
+    }
```
CORS에 대해서는 좀 더 많이 공부해보시면 좋을 것 같아요.
기술면접에서도 자주 나오는 주제기도 하고요.
`Access-Control-Allow-XXX` 헤더들과 그 값을 어떻게 설정할 것인지가 중요하거든요.

kimnayeon0108 replied:
넵 이부분은 오늘 더 공부해보겠습니다! 면접에서도 많이 나오는군요...ㅠ_ㅠ

ksundong commented on BE/src/main/java/baseball/CorsFilter.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.IOException;
+
+@Component
+public class CorsFilter implements Filter {
```
Filter는 Servlet Spec입니다.
그렇다면, Spring에서는 Cors 문제를 해결하기 위한 기능을 제공하지 않고 있을까요?
한 번 알아보시면 좋을 것 같습니다.

ksundong replied:
또한, Filter가 어떤 식으로 동작하는지도 꼭 공부해보시기 바랍니다.

kimnayeon0108 replied:
넵! Filter도 더 자세히 학습해봐야겠네요. 
처음에 스프링에서 cors 해결하는 법을 찾아서 Config클래스(WebMvcConfigurer를 상속받은)에 addCorsMappings를 오버라이딩 하는 방법과 `@CrossOrigin` 어노테이션을 컨트롤러 클래스단에 추가하는 방법 두가지를 적용해보았습니다. 그런데 두가지 방법 모두 아래와 같은 헤더가 추가된 결과가 나왔습니다.
Vary: Origin 
Vary: Access-Control-Request-Method
Vary: Access-Control-Request-Headers

프론트엔드분들에게 이런식으로 해도 되는지 물어봤는데 잘 모르겠다고 해서 더 안전하게(?) 직접 Access-Controll-Allow-Origin 헤더를 추가하는 방식으로 진행했습니다.

Cors 라는 걸 이번에 처음 알게 되었는데 이부분을 더 자세히 학습해야 코드에 해결방법을 적절히 적용할 수 있을 듯 합니다.

ksundong replied:
네 인터넷에 조금 다양한 정보들이 있을 텐데요.
저도 어느 컴퓨터에서는 되고, 어느 컴퓨터에서는 안되는 경우도 있었어서 해결책이 꽤나 중구난방일거에요.
잘 학습해보시면 도움이 될겁니다!

ksundong commented on BE/src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,35 @@
+@RequestMapping("/games")
+public class GameController {
+
+    private GameService gameService;
```
생성자 인젝션을 사용해주는 경우 `final`을 붙이는 것이 낫다고 생각합니다.

kimnayeon0108 replied:
Controller 뿐만 아니라 Service단에서도 Repository 주입하는 부분 final로 수정했습니다!

ksundong commented on BE/src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @GetMapping
+    public List<GameDTO> showGames() {
```
`List`를 그대로 리턴해주는 형태는 가급적이면 API 디자인을 할 때에는 지양해주세요.
breaking Change가 쉽게 발생할 수 있는 형태의 API입니다.

ksundong replied:
이를 해결하기 위해선, 한 번 감싸주는 Wrapping DTO를 만들어주면 좋을 것 같네요.

kimnayeon0108 replied:
GameDTO와 Game의 필드가 똑같아서 GameDTO를 Wrapping DTO의 용도로 바꾸었습니다. 
GameDTO에 Game을 담은 List를 만들었습니다!

ksundong commented on BE/src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    @PostMapping("/{teamId}/score")
+    public void saveScore(@PathVariable Long teamId, @RequestBody RequestScoreDTO scoreDTO) {
```
```suggestion
    public void saveScore(@PathVariable Long teamId, @RequestBody ScoreRequest scoreRequest) {
```
저는 가급적이면 요청과 응답에 대한 DTO를 네이밍 할 때에는 위의 네이밍 전략을 사용하고 있습니다.
DTO라는 불필요한 용어를 제거할 수 있다면 제거하는게 좋다고 생각해요.

kimnayeon0108 replied:
감사합니다! RequestxxxDTO 형태로 작성한 클래스명들 전부다 변경했습니다.

ksundong commented on BE/src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,35 @@
+
+    @PostMapping("/{teamId}/score")
+    public void saveScore(@PathVariable Long teamId, @RequestBody RequestScoreDTO scoreDTO) {
+        gameService.saveScore(teamId, scoreDTO.getInningNumber(), scoreDTO.getScore());
```
```suggestion
        gameService.saveScore(teamId, scoreDTO);
```
getter는 최대한 사용하지 않고, 로직을 진행하면 어떨까요?

kimnayeon0108 replied:
넵 RequestScoreDTO는 ScoreRequest로 이름 변경하였고, 해당 타입 변수만 매개변수로 받도록 수정했습니다!

ksundong commented on BE/src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,35 @@
+        gameService.saveScore(teamId, scoreDTO.getInningNumber(), scoreDTO.getScore());
+    }
+
+    @GetMapping("/{gameId}/scores")
```
조금 API에 대한 설명이 필요한 것 같아요.
- 게임이 무엇인지?
- API 엔드포인트를 보면 `/games/{team_id}/score`, `/games/{game_id}/scores`라는 부분이 존재하는데, 제가 이해하기로는 `{team_id}`는 `{game_id}`의 하위요소라고 생각됩니다. 해당 team의 score를 추가하면 어떤 game인지도 알아야 하지 않을까요?

kimnayeon0108 replied:
야구게임을 시작하면 홈팀 vs 원정팀 의 구조로 5개가 뜨고 플레이어가 이 중 하나를 선택하도록 되어있습니다. 
하나의 홈팀 vs 원정팀 이것을 하나의 game이라고 간주했습니다.
처음에 score를 추가하는 url 설계시 team_id도 넣어야 된다고 생각했는데 구현하다보니 team_id를 따로 사용하지 않게 되어서 없앴습니다.

지금 생각하보니, team이 해당 game에 속하는지 확인하는 작업을 추가하면 좋을거 같아서 코드 수정했습니다!
saveScore 메서드에 매핑된 url도 /games/{gameId}/{teamId}/score 로 변경하였습니다.

ksundong commented on BE/src/main/java/baseball/controller/GameController.java:
```diff
@@ -0,0 +1,35 @@
+
+    @GetMapping("/{gameId}/scores")
+    public GameScoreDTO showScore(@PathVariable Long gameId) {
+        return gameService.convertToGameScoreDTO(gameId);
```
메서드 명이 정확하지 않네요.
`convertTo~` 라면 `gameId`를 가지고 `GameScoreDTO`로 변환해야 할 텐데, 그런 로직이 아닐거라고 생각되거든요.
항상 네이밍에 신경써봅시다. 원래 네이밍은 어려우니까 천천히 생각해보세요.

kimnayeon0108 replied:
맞습니다ㅜㅜ 일단은 떠오르는게 getGameScoreDTO 밖에 없어서 이렇게 변경했습니다. 네이밍은 신경쓴다고 해도 어렵네요ㅠㅠ 더 생각해보겠습니다

ksundong commented on BE/src/main/java/baseball/controller/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,14 @@
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(NotFoundException.class)
+    public String handleNotFoundException(NotFoundException e) {
+        return e.getMessage();
+    }
```
예외처리는 이전 프로젝트에서 사용하던 방식을 계속해서 재사용하면서 업그레이드 시켜봅시다.
좋은 코드는 계속해서 재사용해도 됩니다. 처음부터 새로 개발하실 필요는 없어요.

kimnayeon0108 replied:
엇 죄송한데 이부분을 이해하지 못했습니다😂 이전 프로젝트와 동일하게 작성했는데 혹시 조금만 더 자세히 설명해주실수 있을까요!

ksundong replied:
아 말 그대로, NotFoundException 말고도 여러 예외 상황을 등장시켜서 좀 더 다양한 예외처리를 하는 것을 말씀드린거에요.
이런 코드는 기본 뼈대만 잡아두면 어느 프로젝트에서나 활용할 수 있습니다. 😊

kimnayeon0108 replied:
아하 네넵 다른 예외상황이 발생할 경우도 더 생각해보고 예외처리를 해보도록 하겠습니다!

ksundong commented on BE/src/main/java/baseball/controller/TeamController.java:
```diff
@@ -0,0 +1,38 @@
+
+    @PostMapping("/{teamId}/{memberId}/record")
+    public void addRecord(@PathVariable Long teamId, @PathVariable Long memberId, @RequestBody RequestRecordDTO recordDTO) {
+        teamService.insertRecord(teamId, memberId, recordDTO.getAtBat(), recordDTO.getHit(), recordDTO.getOut());
```
여기도 일일이 필드를 꺼내주기 보다는, 내부에서 처리하는 형태로 구현해주세요.

kimnayeon0108 replied:
넵 이부분도 RecordRequest( RequestRecordDTO 에서 이름이 변경된 클래스 )타입의 변수만 매개변수로 받도록 변경하고, RecordRequest 내부에 Record를 생성해주는 메소드를 추가했습니다~

ksundong commented on BE/src/main/java/baseball/domain/Member.java:
```diff
@@ -0,0 +1,43 @@
+        this.record = record;
+    }
+
+    public boolean hasRecord() {
+        return record != null;
+    }
```
👍

ksundong commented on BE/src/main/java/baseball/domain/Record.java:
```diff
@@ -0,0 +1,36 @@
+
+public class Record {
+
+    private int atBat;
```
이 필드의 의미가 무엇인가요?

kimnayeon0108 replied:
타석 입니다!

ksundong commented on BE/src/main/java/baseball/domain/Record.java:
```diff
@@ -0,0 +1,36 @@
+    private int atBat;
+    private int hit;
+    private int out;
+    private double average;
```
사실 이 필드는 유도될 수 있다고 봅니다.
과연 상태로 관리가 되어야 할까요?
오히려 이 필드를 매번 계산해줘야 할 텐데, 비효율적으로 보입니다.

ksundong replied:
- 한번 average를 제거하는 방향도 고려해보세요!

kimnayeon0108 replied:
그럼 average를 DB에 저장하지 않고, 프론트단에서 계산하도록 하는게 더 좋은 방향일까요??

ksundong replied:
프론트에서 계산해줄 필요 없이, domain이 직접 계산해주면 되지 않을까요?

ksundong replied:
```java
public double getAverage() {
    return (double) hit / atBat;
}
```
으로 계산해주면 될 것 같은데요~

kimnayeon0108 replied:
아하... 왜 이런생각을 못했을까요 🤯 프론트엔드분들이 API 수정 요청한 부분에서 average를 빼달라고 해서 더이상은 이 메소드가 필요없어졌습니다. 코드는 지우고 머리속에 저장해두겠습니다~~

ksundong commented on BE/src/main/java/baseball/domain/Record.java:
```diff
@@ -0,0 +1,36 @@
+        return atBat / (double) hit;
+    }
+
+    public int getAtBat() {
+        return atBat;
+    }
+
+    public int getHit() {
+        return hit;
+    }
+
+    public int getOut() {
+        return out;
+    }
+
+    public double getAverage() {
+        return average;
+    }
```
getter를 무조건 선언하기 보다는 필요한 시점에 선언해주세요.
Domain 객체에 getter는 꼭 필요한 시점에 추가가 되는것이 객체지향 프로그래밍을 하는데 도움이 됩니다.

kimnayeon0108 replied:
앗 네 이부분도 더 신경쓰겠습니다.

kimnayeon0108 replied:
사용되지 않는 Domain 들의 getter는 삭제했습니다. 이부분 기억하고 유의해서 프로그래밍 해야겠어요😖

ksundong commented on BE/src/main/java/baseball/domain/Record.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private double calculateAverage(int atBat, int hit) {
+        return atBat / (double) hit;
```
형변환이 어디서 이루어지는지 명확하네요.
형변환의 연산 순서도 한 번 생각해보시면
`(double) atBat / hit` 도 가능함을 알 수 있습니다.

ksundong commented on BE/src/main/java/baseball/domain/Score.java:
```diff
@@ -0,0 +1,20 @@
+package baseball.domain;
+
+public class Score {
```
Set에서의 중복방지를 위해서는 equals & hashCode를 재정의 할 필요성도 있는 것 같네요.

kimnayeon0108 replied:
이부분까지는 생각을 못했네요ㅠㅠ  Score와 더불어 Member들도 Set으로 관리되고 있어서 Member에도 equals와 hashCode 추가했습니다!

ksundong commented on BE/src/main/java/baseball/domain/Team.java:
```diff
@@ -0,0 +1,49 @@
+    private Long id;
+
+    private String name;
+    private Set<Member> members;
+    private Set<Score> scores;
```
이 둘은 빈 컬렉션으로 초기화가 될 수 있도록 방어로직을 구현해주세요.
자칫 잘못하면 NPE가 발생할 수 있습니다.

kimnayeon0108 replied:
scores 같은 경우에는 생성자로 값이 주어지는데 이경우에도 빈 컬렉션으로 초기화를 해주어야할까요? 필드에서 초기화를 하려고 하니 IDE에서 회색으로 비활성화 되어서 적어주나마나 인거 같아서 질문드립니다..

members는 필드에서 바로 new HashSet<>(); 으로 초기화 시켜줬습니다.

ksundong replied:
인자가 null인 경우도 고려되었나요?

kimnayeon0108 replied:
앗 이부분 scores랑 members를 반대로 이야기 했네요

scores가 아닌 members가 생성자로 값이 주어지고,
members가 아닌 scores를 필드에서 바로 초기화 해줬습니다.

isMemberEmpty 라는 메서드를 추가해줘서 인자로 받아온 members 값이 null인 경우 예외를 발생하도록 코드 수정했습니다~

ksundong commented on BE/src/main/java/baseball/domain/Team.java:
```diff
@@ -0,0 +1,49 @@
+
+    public void setScore(Score score) {
+        scores.add(score);
+    }
```
`setXXX` 보다 좋은 네이밍을 고민해봐주세요.
`scores`는 `null`일 수 있을 것 같네요.

kimnayeon0108 replied:
addScore로 변경했습니다~

ksundong commented on BE/src/main/java/baseball/domain/Team.java:
```diff
@@ -0,0 +1,49 @@
+        scores.add(score);
+    }
+
+    public Member getMemberById(Long id) {
+        for (Member member : members) {
+            if (member.getId() == id) {
+                return member;
+            }
+        }
+        throw new NullPointerException();
+    }
```
`stream`을 활용해주세요.
`Optional`을 리턴해줘도 좋을 것 같고, 이곳에서 직접 예외처리를 해줄 수도 있겠네요.
`NPE`를 직접 던지기보다는, Custom 예외를 지정해서 발생시켜주세요.

kimnayeon0108 replied:
이부분 스트림으로 변경했습니다. 그리고 MemberNotFoundException 이라는 사용자 정의 예외를 생성해서 발생시켰습니다~

ksundong commented on BE/src/main/java/baseball/exception/GameNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+public class GameNotFoundException extends NotFoundException {
+
+    private static final String MESSAGE = "game이 존재하지 않습니다.";
```
MESSAGE로 퉁치기 보다는 명확한 네이밍을 지정해주세요.
가급적이면 저는 이런 message의 경우엔 테스트 할 때, 필요해서 public으로 지정해주는 편입니다.

kimnayeon0108 replied:
커스텀 예외에 정의된 메시지들에 의미가 담기도록 모두 수정했습니다!

kimnayeon0108 replied:
이번 미션은 테스트 코드를 작성하지 않고 진행했는데 다음부터는 꼭 테스트 코드를 작성 해야겠다는 생각이 듭니다
Dion의 피드백을 보니 제 코드에 문제가 많은데 테스트 코드를 작성했더라면 좀 더 꼼꼼히 구현할 수 있었을텐데 라는 생각이 절실하게 드네요.ㅠㅠ

ksundong commented on BE/src/main/java/baseball/exception/RecordDTONotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+package baseball.exception;
+
+public class RecordDTONotFoundException extends NotFoundException {
```
`DTO`는 제거해주세요~

kimnayeon0108 replied:
RecordDTO라는 클래스를 삭제하게 되어서 이 예외도 삭제해주었습니다!

ksundong commented on BE/src/main/java/baseball/repository/TeamRepository.java:
```diff
@@ -0,0 +1,14 @@
+public interface TeamRepository extends CrudRepository<Team, Long> {
+
+    @Query("select at_bat, hit, `out`, average, name, position from record left join member on record.member = id where id = :id")
+    Optional<RecordDTO> findRecordByMemberId(Long id);
```
리턴타입을 `RecordDTO`로 주는 것은 저는 별로 권장하기 힘든 형태네요.
실제로 IDE에서도 밑줄이 그어집니다. `Record`가 필요하다면 `RecordRepository`를 이용해야하지 않나 싶어요.

그리고 Join을 할 때에는 해당 필드가 어떤 entity의 필드인지 명시해주면 더 좋을 것 같습니다.

kimnayeon0108 replied:
아직 지식이 부족하지만, Spring Data JDBC의 aggregate 라는 개념의 root에 해당하는 Repository만 만들어서 구현하는게 제가 이해한 컨셉이여서 시도해보았습니다...! 따라서 RecordRepository를 만들지 않았는데 제대로 이해하지 못한 상태로 구현한거 같아요

Join을 할때, 이 경우 member와 record의 속성을 전부 가지고 있는 클래스가 필요한데 이럴때는 클래스명을 어떤식으로 짓는것이 좋을지 궁금합니다..!

ksundong replied:
aggregate root는 어쩌면 Team이 아닐까요?
Repository는 데이터베이스와 연결부인 DAO일 뿐이구요.
그렇다면 Repository는 여러개가 있어도 괜찮지 않을까요?
저도 DDD 컨셉에 대한 이해가 부족해서, 좋은 리뷰를 드리지 못할 것 같네요.

Join을 해서 둘 다 가지고 있는 클래스라면, 왜 그게 필요한건지 먼저 생각해볼 것 같아요.
저는 그냥 MemberRecord로 빨리 지어주고, 나중에 좋은 이름을 구현하면서 다시 짓는 편입니다!

kimnayeon0108 replied:
아하! 공식문서를 봐도 아직도 어려운거 같습니다.... 답변 감사합니다~

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public List<GameDTO> getGameDTOList() {
+        if (gameRepository.count() != 5) {
```
5는 상수죠? 추출해봅시다.

ksundong replied:
당장 이 코드만 봐서는 왜 5가 아니면 저장하는지 잘 모르겠네요.

kimnayeon0108 replied:
EMPTY라는 상수를 만들고 0이라는 값을 줬습니다.
그리고 gameReopitory.count() == EMPTY 이면 game을 saveGame을 호출하도록 코드 수정했습니다~

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        List<GameDTO> gameDTOs = new ArrayList<>();
+        Iterable<Game> games = gameRepository.findAll();
```
`games`는 inline해주는 편이 더 좋아보이네요.

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,78 @@
+            saveGames();
+        }
+
+        List<GameDTO> gameDTOs = new ArrayList<>();
+        Iterable<Game> games = gameRepository.findAll();
+
+        for (Game game : games) {
+            Long id = game.getId();
+            Team homeTeam = teamRepository.findById(game.getHomeTeamId()).orElseThrow(TeamNotFoundException::new);
+            Team awayTeam = teamRepository.findById(game.getAwayTeamId()).orElseThrow(TeamNotFoundException::new);
+
+            GameDTO gameDTO = new GameDTO(id, homeTeam.getName(), awayTeam.getName());
+            gameDTOs.add(gameDTO);
+        }
+        return gameDTOs;
```
```suggestion
    return StreamSupport.stream(gameRepository.findAll().spliterator(), false)
        .map(game -> {
          Long id = game.getId();
          Team homeTeam = teamRepository.findById(game.getHomeTeamId()).orElseThrow(TeamNotFoundException::new);
          Team awayTeam = teamRepository.findById(game.getAwayTeamId()).orElseThrow(TeamNotFoundException::new);

          return new GameDTO(id, homeTeam.getName(), awayTeam.getName());
        })
        .collect(Collectors.toList());
```
Stream을 이용하면 이런식으로도 되겠죠?

kimnayeon0108 replied:
아직 스트림 사용이 익숙치 않은듯 합니다... GameDTO가 내부적으로 List<Game>을 갖도록 변경되어서 이부분도 현재는 필요하지 않아서 다른 부분 수정할때 적용해보겠습니다~

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,78 @@
+    private void saveGames() {
+        List<Game> games = new ArrayList<>();
+
+        for (long i = 1; i <= NUMBER_OF_TEAM; i += 2) {
```
for문은 너무 복잡하면 안됩니다.
가독성이 현격히 떨어지게됩니다.

ksundong replied:
팀의 수가 10인건가요?

kimnayeon0108 replied:
네 처음에 DB에 저장되는 team의 수가 10입니다.

kimnayeon0108 replied:
이부분 어떻게 개선하면 좋을지 더 고민해보겠습니다!

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,78 @@
+        teamRepository.save(team);
+    }
+
+    public GameScoreDTO convertToGameScoreDTO(Long gameId) {
+        Game game = gameRepository.findById(gameId).orElseThrow(GameNotFoundException::new);
+        Team homeTeam = teamRepository.findById(game.getHomeTeamId()).orElseThrow(TeamNotFoundException::new);
+        Team awayTeam = teamRepository.findById(game.getAwayTeamId()).orElseThrow(TeamNotFoundException::new);
+
+        return new GameScoreDTO(gameId, homeTeam, awayTeam);
+    }
```
조회로직인데, convert라는 동사는 전혀 맞지 않네요.

kimnayeon0108 replied:
getGameScoreDTO 라고 변경했습니다~

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,78 @@
+        Team team = teamRepository.findById(teamId).orElseThrow(TeamNotFoundException::new);
+
+        Score newScore = new Score(inningNumber, score);
+        team.setScore(newScore);
```
`addScore`가 더 좋을 것 같네요.

kimnayeon0108 replied:
네넵 addScore라고 변경했습니다~!

ksundong commented on BE/src/main/java/baseball/service/TeamService.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public List<TeamDTO> convertToTeamDTOList() {
+        Iterable<Team> teams = teamRepository.findAll();
+        List<TeamDTO> teamDTOs = new ArrayList<>();
+
+        for (Team team : teams) {
+            Set<Member> members = team.getMembers();
+            TeamDTO teamDTO = new TeamDTO(team, members);
+            teamDTOs.add(teamDTO);
+        }
+        return teamDTOs;
```
여기도 stream을 이용해서 개선해봅시다.

kimnayeon0108 replied:
stream으로 변경했습니다~!

ksundong commented on BE/src/main/java/baseball/service/TeamService.java:
```diff
@@ -0,0 +1,67 @@
+        Team team = teamRepository.findById(teamId).orElseThrow(TeamNotFoundException::new);
+        Set<Member> members = team.getMembers();
+
+        Set<RecordDTO> recordDTOS = new HashSet<>();
+        for (Member member : members) {
+            if (member.hasRecord()) {
+                RecordDTO recordDTO = teamRepository.findRecordByMemberId(member.getId())
+                        .orElseThrow(RecordDTONotFoundException::new);
+                recordDTOS.add(recordDTO);
+            }
+        }
```
되도록이면 이런 로직은 stream을 이용하는게 가독성 면에서 좋을 것 같아요.

kimnayeon0108 replied:
이부분 프론트엔드분들이 game단위로 반환해달라고 하셔서 GameService 쪽으로 로직 옮겼습니다!
그리고 stream으로 변경도 완료했습니다~

ksundong commented on BE/src/main/java/baseball/service/TeamService.java:
```diff
@@ -0,0 +1,67 @@
+        Team team = teamRepository.findById(teamId).orElseThrow(TeamNotFoundException::new);
+        Set<Member> members = team.getMembers();
+
+        Set<RecordDTO> recordDTOS = new HashSet<>();
```
여기는 DTOS네요.
다른 곳은 DTOs였습니다.

kimnayeon0108 replied:
DTOs 로 통일했습니다~!

ksundong commented on BE/src/main/java/baseball/service/dto/GameDTO.java:
```diff
@@ -0,0 +1,27 @@
+    private String homeTeamName;
+    private String awayTeamName;
+
+    public GameDTO(Long id, String homeTeamName, String awayTeamName) {
+        this.id = id;
+        this.homeTeamName = homeTeamName;
+        this.awayTeamName = awayTeamName;
+    }
```
Game을 인자로 받아서 생성하는 것도 생각해보세요~

ksundong commented on BE/src/main/java/baseball/service/dto/GameScoreDTO.java:
```diff
@@ -0,0 +1,36 @@
+        teamScoreDTOS.add(homeTeamScoreDTO);
+        teamScoreDTOS.add(awayTeamScoreDTO);
+
+        return teamScoreDTOS;
```
왜 List형태로 관리해주는지 의문이에요.
각각 필드로 지정해주는게 더 나을 것 같다는 생각이 듭니다.

kimnayeon0108 replied:
맞습니다... home이랑 away 각각 따로 분리해서 필드 갖도록 변경했습니다~

ksundong commented on BE/src/main/resources/application.properties:
```diff
@@ -0,0 +1,6 @@
+spring.datasource.url=jdbc:mysql://10.1.11.205:3306/baseball?autoReconnect=true&serverTimezone=Asia/Seoul&characterEncoding=UTF-8
```
DB 접속정보는 노출이 되지 않도록 해주시는게 좋을 것 같네요.

kimnayeon0108 replied:
앗! 감사합니다 localhost로 수정했습니다.

ksundong reviewed (changes requested):
안녕하세요! yeon! 리뷰어 Dion입니다. 미션 수행하느라 고생이 많으십니다.
야구 미션은 미션 자체가 굉장히 자유도도 높고, 여러가지 생각할 거리들이 많기 때문에 많은 것을 고민하게 되는 프로젝트입니다.
yeon의 조는 백엔드가 게임을 처리하는 것보다 데이터를 관리하는 역할의 백엔드로 보여지네요.

적절한 시기에 리뷰요청을 보내주신 것 같아요.
코드도 직접 클론 받아서 읽고 개선해보면서 리뷰해보았습니다.

한 번 생각해보셨으면 하는 부분은 다음과 같아요.
- stream의 활용을 좀 더 고민해봐주세요.
- 네이밍에 대해서 좀 더 고민해봐주세요.
- 예외 처리기를 좀 더 발전시켜서 계속해서 써먹을 수 있도록 해주세요.
- getter를 사용하는 것을 지양해주시고, 불필요하게 getter를 미리 선언해주기 보다는, 필요한 시점에만 선언해주세요.
- REST API를 작성하는 것은 어렵지만, 잘 익혀두어야 하는 부분 중 하나입니다. 조금 더 신경써주세요.
- `null`을 항상 조심해주시고, `NPE`가 발생하지 않는 코딩에 대해서도 고민해주세요.
- 다른 사람이 읽었을 때, 잘 읽히는 코드가 좋은 코드라고 생각해주시고, 코딩하시다보면 개선될 수 있는 부분이 보이실 것 같습니다.

세부적인 내용은 코멘트 참고해주시고, 궁금하시거나 도움이 필요한 부분이 있으면 언제든 코멘트 남겨주세요.
고생하셨습니다!

kimnayeon0108 committed:
Feat: update application.properties

kimnayeon0108 committed:
Refactor: Fix GameDTO to contain game List in order to avoid return List type in Controller

kimnayeon0108 committed:
Refactor: Fix DTO class name regarding Request

kimnayeon0108 committed:
Refactor: add get Score object with ScoreRequest object

kimnayeon0108 committed:
Feat: fix API URL and add validation logic

fix URL when inserting score data with POST type HTTP method, /games/{teamId}/score to /games/{gameId}/{teamId}/score.
add validation logic to check if Team is in corresponding Game

kimnayeon0108 committed:
Refactor: change method name, convertToGameScoreDTO to getGameScoreDTO

kimnayeon0108 committed:
Refactor: clarify exception message name

kimnayeon0108 committed:
Refactor: remove useless getter

kimnayeon0108 committed:
Feat: add equals and hashCode method

kimnayeon0108 committed:
Feat: add equals and hashCode method

kimnayeon0108 committed:
Feat: add initializing Set<Score>, MemberNotFoundException
Refactor: change method name, setScore to addScore

kimnayeon0108 committed:
Refactor: fix import statement

kimnayeon0108 committed:
Refactor: add GameMemberDTO

kimnayeon0108 committed:
Feat: add Delete all data when restarting game

kimnayeon0108 committed:
Refactor: change the way to create id of Game entity

kimnayeon0108 committed:
Feat: fix GET type URL when showing records of members by game and remove useless method

kimnayeon0108 committed:
Merge branch 'feature/BE/deploy' into dev/BE

kimnayeon0108 committed:
Feat: remove average field considered to be useless

kimnayeon0108 committed:
Refactor: add EMPTY constant meaning zero, add JsonIgnore annotation, fix getGameDTOList method name to getGameDTO and fix getMemberById method as stream

kimnayeon0108 committed:
Refactor: remove useless custom exception

kimnayeon0108 committed:
Refactor: remove useless method

kimnayeon0108 committed:
Refactor: fix getGameTeamDTO method as stream and separate method named getRecordDTOList

kimnayeon0108 committed:
Chore

kimnayeon0108 committed:
Refactor: separate method finding object from repository

kimnayeon0108 committed:
Feat: add TeamsDTO for wrapping TeamDTO and fix getting TeamsDTO logic as stream

kimnayeon0108 committed:
Feat: add TeamsDTO for wrapping TeamDTO and fix getting TeamsDTO logic as stream

kimnayeon0108 committed:
Refactor: fix deleteGame method as stream

kimnayeon0108 committed:
Refactor: fix GameDTO constructor parameter type to List

kimnayeon0108 committed:
Refactor: fix GameDTO, TeamDTO and TeamScoreDTO constructor parameter type

kimnayeon0108 committed:
Refactor: fix GameScoreDTO field

kimnayeon0108 committed:
Feat: add validation check logic

kimnayeon0108 commented:
@ksundong Dion 수요일에 피드백 달아주셨는데 API 수정 요청을 받아서 수정하느라 피드백 수정을 늦게 하게 되었습니다. 오랫동안 질질 끈거같아서 죄송해요.. 피드백 보면서 네이밍과 stream 사용부분에 특히 더 신경써야겠다는 생각이 들었습니다.
피드백 감사합니다~!

ksundong commented on BE/src/main/java/baseball/domain/Game.java:
```diff
@@ -0,0 +1,48 @@
+    private Long awayTeamId;
+
+    @Transient
+    @JsonIgnore
```
JsonIgnore는 왜 붙여주신건가요?

kimnayeon0108 replied:
아래 설명대로 Persistable을 implements해서 isNew 라는 메소드를 오버라이딩하고 전역변수로도 갖고 있습니다. 
그런데 JSON 타입을 반환될 때 "new"라는 속성이 추가되어서 JsonIgnore 어노테이션을 붙였습니다~

ksundong replied:
생각해보면, 도메인이 그대로 반환된다는 건데, 이는 지양해주시는게 좋을 것 같아요.
DTO를 사용해보면 어떨까요?

kimnayeon0108 replied:
Games라는 DTO클래스로 한번 감싸기 때문에 도메인을 그대로 사용하는게 더 낫지 않을까 라는 생각을 했습니다. DTO클래스끼리 계속 중첩되는 느낌이 들어서요! 그래도 DTO를 사용하는게 더 좋다고 이해하면 될까요~??

GameDTO라는 클래스를 생성해서 JSON 데이터에 필요한 속성들만 전역으로 두도록 변경했습니다~

ksundong replied:
도메인의 역할은 데이터를 전달하는 것이 아니니까요~
DTO의 필드와 도메인의 필드가 1:1로 대응할 필요도 없습니다.

ksundong commented on BE/src/main/java/baseball/domain/Game.java:
```diff
@@ -0,0 +1,48 @@
+import org.springframework.data.annotation.Transient;
+import org.springframework.data.domain.Persistable;
+
+public class Game implements Persistable<Long> {
```
👍

ksundong replied:
어떻게 찾으셨는지 궁금하네요 ㅎㅎ

kimnayeon0108 replied:
야구게임을 종료할때 클라이언트에서 Delete 요청을 보내면 game, team의 score, 선수들의 record를 삭제하도록 기능을 추가했습니다.
게임을 다시 시작하고 새로운 game 객체를 생성하고, DB에 담을 경우 game 인스턴스의 id값이 6부터 들어가는 문제가 발생해서 id값을 직접 지정해주는 방법을 찾다가 Persistable을 사용하는 방법을 찾았습니다!

kimnayeon0108 replied:
https://skyblue300a.tistory.com/9
이 글을 참고하여 Persistable을 사용했습니다!

sanhee replied:
오 저도 이방법 나중에 써봐야겠네요. 감사합니다!

ksundong commented on BE/src/main/java/baseball/domain/Team.java:
```diff
@@ -0,0 +1,62 @@
+        scores.clear();
+    }
+
+    private boolean isMembersEmpty(Set<Member> members) {
+        if (!members.isEmpty()) {
+            throw new MemberNotFoundException();
+        }
+        return false;
+    }
```
void 형태라면 return false가 없어져도 될거에요.

ksundong replied:
그리고 private 메서드는 사용되는 지점 아래에 작성해주는 것이 가독성에 좋습니다.

kimnayeon0108 replied:
void로 변경하면서 validateMembers라는 이름으로 메소드명 변경하였고, 호출되는 생성자 바로 아래로 위치 변경했습니다!

ksundong commented on BE/src/main/java/baseball/exception/MemberNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+public class MemberNotFoundException extends NotFoundException {
+
+    private static final String MEMBER_NOT_FOUND_MESSAGE = "member가 존재하지 않습니다.";
```
MESSAGE를 적어주지 않아도 괜찮다고 생각해요.

kimnayeon0108 replied:
그렇군요... 커스텀 예외에서 메세지에 해당하는 부분 수정했습니다~

ksundong commented on BE/src/main/java/baseball/repository/GameRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface GameRepository extends CrudRepository<Game, Long> {
+
+    List<Game> findAll();
```
Override 애노테이션이 누락되었네요.

kimnayeon0108 replied:
오호... 굳이 명시하지 않아도 된다고 생각했는데 아니였군요... Override 애노테이션 추가했습니다~!

ksundong replied:
가급적이면 명시적으로 달 수 있는경우 달아주는 것이 좋습니다.

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,116 @@
+        if (gameRepository.count() == EMPTY) {
+            saveGames();
+        }
+        List<Game> games = gameRepository.findAll();
```
Paging도 고려해주세요!

ksundong commented on BE/src/main/java/baseball/service/GameService.java:
```diff
@@ -0,0 +1,116 @@
+    public void deleteGame() {
+        gameRepository.deleteAll();
+
+        List<Team> teams = teamRepository.findAll().stream()
+                .map(team -> {
+                    team.deleteScores();
+                    team.getMembers().stream()
+                            .forEach(Member::deleteRecord);
+                    return team;
+                }).collect(Collectors.toList());
```
team에게 메시지를 전달하는 형태로 구현해볼까요?

kimnayeon0108 replied:
members의 스트림을 생성해서 사용하는 대신, Team에 deleteRecordOfMembers라는 메소드를 추가해서 이를 호출하도록 변경하였습니다~

ksundong commented on BE/src/main/java/baseball/service/TeamService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public TeamsDTO getTeamsDTO() {
+        List<TeamDTO> teamDTOs = teamRepository.findAll().stream()
+                .map(team -> new TeamDTO(team))
```
```suggestion
                .map(TeamDTO::new)
```

ksundong commented on BE/src/main/java/baseball/service/dto/GameDTO.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+public class GameDTO {
```
Games가 좋지 않을까요?

kimnayeon0108 replied:
Games로 변경하였습니다! 다음 미션때는 DTO관련 클래스명에 관해 더 신경써서 진행 해야겠습니다ㅜ_ㅜ

ksundong reviewed (approved):
피드백 반영하느라 고생하셨습니다!
머지는 일요일 밤에 할게요!

kimnayeon0108 committed:
Refactor: fix method to validate members

kimnayeon0108 committed:
Refactor: fix message constant name

kimnayeon0108 committed:
Refactor: add missing override annotation

kimnayeon0108 committed:
Refactor: fix deleteGame method

kimnayeon0108 committed:
Refactor: fix getTeamsDTO

kimnayeon0108 committed:
Refactor: fix class name GameDTO to Games

kimnayeon0108 committed:
Refactor: remove useless column in schema

kimnayeon0108 committed:
Feat: add GameDTO class
